### PR TITLE
Add projected service account support

### LINKED ISSUE #78: Support projected ServiceAccount volumes
The current implementation uses the workloads normal ServiceAccount token at `/var/run/secrets/kubernetes.io/serviceaccount/token`. This token does not expire and allows access to the Kubernetes API allowing Vault to impersonate the workload.

Support for [projected ServiceAccount volumes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) would allow the token to be rotated and would mean that if the token were intercepted it could only be used to authenticate to Vault (rather than have wider impacts against the Kubernetes API).

### PR DESCRIPTION
**Pull request by tvoran**
Continues #79. Fixes #78.

Adds more unit tests for `serviceaccount()`, and reads the tokenPath from the volumeMount for projected tokens.

### Description
This PR adds the ability to specify a [projected service token](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) for vault-agent to use with auto-auth against the vault kubernetes auth method.

It adds a new annotation, `vault.hashicorp.com/agent-service-account-token-volume-name`, which is the name of the projected volume holding the token. If the volume is mounted to another container in the deployment, the token volume will be mounted to the same location in the vault-agent containers. Otherwise it will be mounted at the default location of `/var/run/secrets/vault.hashicorp.com/serviceaccount/`.

### Example
Take the [example nginx Pod](https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/pods/pod-projected-svc-token.yaml) from [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection), and add these annotations:

```yaml
  annotations:
    # AGENT INJECTOR SETTINGS
    vault.hashicorp.com/agent-inject: "true"
    vault.hashicorp.com/agent-inject-status: "update"
    vault.hashicorp.com/agent-inject-secret-kv-secret: "internal/nginx"
    vault.hashicorp.com/agent-inject-template-kv-secret: |
      {{ with secret "internal/nginx" }}{{ .Data | toJSONPretty }}{{ end }}
    # VAULT SETTINGS
    vault.hashicorp.com/role: "app"
    vault.hashicorp.com/agent-service-account-token-volume-name: "vault-token"
```

Configure vault to accept the token with a role something like this:

```shell
vault write auth/kubernetes/role/app \
      bound_service_account_names=build-robot \
      bound_service_account_namespaces=default \
      audience=vault \
      policies=internal-app \
      ttl=24h
```

Once the Pod is deployed, vault-agent will be configured to use the token `/var/run/secrets/tokens/vault-token` for auto-auth. That token is periodically refreshed by kubernetes, and vault-agent will read the new token and use it whenever its refreshed.

One nice thing about this setup is that since the vault role is configured to only allow tokens that have an audience of `vault`, only the projected token will work, since the normal service token will have an audience of something like `https://kubernetes.default.svc.cluster.local`.

And in this example, to mount the projected token **only** on the vault-agent containers, remove the `volumeMounts` block from the nginx container. Then vault-agent will be configured to use the token at `/var/run/secrets/vault.hashicorp.com/serviceaccount/vault-token` for auto-auth.

### CODE REVIEW

File: agent-inject/agent/annotations.go [hunk 1]
```diff
@@ -146,6 +147,10 @@ const (
 	// and gid) is set on the injected Vault Agent containers
 	AnnotationAgentSetSecurityContext = "vault.hashicorp.com/agent-set-security-context"
 
+	// AnnotationAgentServiceAccountTokenVolumeName is the optional name of a volume containing a
+	// service account token
+	AnnotationAgentServiceAccountTokenVolumeName = "vault.hashicorp.com/agent-service-account-token-volume-name"
```
Comment (reegnz) [inline agent-inject/agent/annotations.go:152]: Please also allow defining the audience for the token with something like:
`vault.hashicorp.com/agent-service-account-token-audience`

Comment (tvoran) [reply agent-inject/agent/annotations.go:152]: This PR is providing the ability to mount and use a projected token, but it's not defining the token itself. So I don't think setting the audience for the token is in scope for this PR.

File: agent-inject/agent/agent.go [hunk 2]
```diff
@@ -629,16 +640,77 @@ func (a *Agent) Validate() error {
...
-	var serviceAccountName, serviceAccountPath string
+func serviceaccount(pod *corev1.Pod) (*ServiceAccountTokenVolume, error) {
+	if volumeName := pod.ObjectMeta.Annotations[AnnotationAgentServiceAccountTokenVolumeName]; volumeName != "" {
+		// Attempt to find existing mount point of named volume and copy mount path
+		for _, container := range pod.Spec.Containers {
+			for _, volumeMount := range container.VolumeMounts {
+				if volumeMount.Name == volumeName {
+					tokenPath := getProjectedTokenPath(pod, volumeName)
+					if len(tokenPath) == 0 {
+						return nil, fmt.Errorf("failed to find tokenPath for service account volume mount %q", volumeName)
```
[the remark this answers is not in the document]
Comment (tvoran) [reply agent-inject/agent/agent.go]: Yep, I agree, that's more idiomatic. Changed in 94b8263.

File: agent-inject/agent/agent.go [hunk 3]
```diff
@@ -629,16 +640,77 @@ func (a *Agent) Validate() error {
...
+						TokenPath: tokenPath,
+					}, nil
+				}
+			}
+		}
+
+		// Otherwise, check the volume exists and fallback to `DefaultServiceAccountMount`
+		for _, volume := range pod.Spec.Volumes {
+			if volume.Name == volumeName {
+				tokenPath := getTokenPathFromProjectedVolume(volume)
```
[the remark this answers is not in the document]
Comment (tvoran) [reply agent-inject/agent/agent.go]: Updated in 94b8263.

File: agent-inject/agent/agent.go [hunk 4]
```diff
@@ -600,8 +609,10 @@ func (a *Agent) Validate() error {
 		return errors.New("namespace missing from request")
 	}
 
-	if a.ServiceAccountName == "" || a.ServiceAccountPath == "" {
-		return errors.New("no service account name or path found")
+	if a.ServiceAccountTokenVolume.Name == "" ||
```
[the remark this answers is not in the document]
Comment (tvoran) [reply agent-inject/agent/agent.go]: Good catch, added in 94b8263.

**Review by benashz — APPROVED** [review]
LGTM!

**Pull request merged**